core/client/src/main/java/org/phenotips/remote/client/internal/DefaultRemotePatientMatchResource.java

```diff
@@ -0,0 +1,214 @@
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.container.Container;
+import org.xwiki.container.Request;
+import org.xwiki.rest.XWikiResource;
+
+import java.util.List;
+
+import javax.annotation.Nonnull;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+import org.apache.commons.lang3.BooleanUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.math.NumberUtils;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+
+/**
+ * Default implementation of the {@link RemotePatientMatchResource}.
+ *
+ * @version $Id$
+ * @since 1.2
+ */
+@Component
+@Named("org.phenotips.remote.client.internal.DefaultRemotePatientMatchResource")
+@Singleton
+public class DefaultRemotePatientMatchResource extends XWikiResource implements RemotePatientMatchResource
+{
+    private static final String REQ_NO = "reqNo";
+
+    private static final String OFFSET = "offset";
+
+    private static final String LIMIT = "maxResults";
+
+    private static final String SERVER = "server";
+
+    private static final String SEND_NEW_REQUEST = "sendNewRequest";
+
+    /** The logging object. */
+    @Inject
+    private Logger logger;
+
+    /** The secure patient repository. */
+    @Inject
+    @Named("secure")
+    private PatientRepository repository;
+
+    /** The remote similar patients matching service. */
+    @Inject
+    private RemoteMatchingService matchingService;
+
+    /** The XWiki container. */
+    @Inject
+    private Container container;
+
+    @Override
+    public Response findRemoteMatchingPatients(final String patientId)
+    {
+        // Get the request container.
+        final Request request = this.container.getRequest();
+        // The patient ID must not be blank.
+        if (StringUtils.isBlank(patientId)) {
+            this.logger.error("Patient ID is not specified.");
+            return Response.status(Response.Status.BAD_REQUEST).build();
+        }
+        // The server must not be blank.
+        final String server = (String) request.getProperty(SERVER);
+        if (StringUtils.isBlank(server)) {
+            this.logger.error("Server is not specified.");
+            return Response.status(Response.Status.BAD_REQUEST).build();
+        }
+        // Get the other parameters, if specified, or set the defaults.
+        final int offset =  NumberUtils.toInt((String) request.getProperty(OFFSET), 1);
+        if (offset < 1) {
+            this.logger.error("The requested offset is out of bounds: {}", offset);
+            return Response.status(Response.Status.BAD_REQUEST).build();
+        }
+        final boolean newRequest = BooleanUtils.toBoolean((String) request.getProperty(SEND_NEW_REQUEST));
+        final int limit = NumberUtils.toInt((String) request.getProperty(LIMIT), -1);
+        final int reqNo = NumberUtils.toInt((String) request.getProperty(REQ_NO), 1);
```

Comment:
Just something to consider: if a UI sends a request with `newRequest == true`, a new request will be sent, which is expected. However if a second page of results is requested, a new request should not be sent. Need to check how UI handles this